src/replica/replica.cpp

```diff
@@ -595,6 +597,35 @@ decree replica::last_prepared_decree() const
     return start;
 }
 
+std::string replica::get_progress_message() const
+{
+    auto message = fmt::format("max_prepared_decree={}, max_plog_decree={}, "
+                               "last_committed_decree={}, last_applied_decree={}, "
+                               "last_flushed_decree={}, last_durable_decree={}, "
+                               "max_gc_decree={}",
+                               max_prepared_decree(),
+                               _private_log->max_decree(get_gpid()),
+                               last_committed_decree(),
+                               last_applied_decree(),
+                               _app->last_flushed_decree(),
+                               last_durable_decree(),
+                               _private_log->max_gced_decree(get_gpid()));
+
+    const auto &dup_messages = _duplication_mgr->get_progress_messages();
+    if (dup_messages.empty()) {
+        return message;
+    }
+
+    message += "\n        duplications {";
```

Comment:
Would it be more appropriate if we return a JSON format string, then the thirdparty tools are convenient to parse the result?